android/java/org/chromium/chrome/browser/playlist/hls_content/HlsUtils.java

```diff
@@ -0,0 +1,187 @@
+            PlaylistStreamingObserverImpl
+                    .PlaylistStreamingObserverImplDelegate playlistStreamingObserverImplDelegate =
+                    new PlaylistStreamingObserverImpl.PlaylistStreamingObserverImplDelegate() {
+                        @Override
+                        public void onResponseStarted(String url, long contentLength) {
+                            // unused for manifest file
+                        }
+
+                        @Override
+                        public void onDataReceived(byte[] response) {
+                            try {
+                                MediaUtils.writeToFile(response, mediaPath);
+                            } catch (Exception ex) {
+                                ex.printStackTrace();
+                            }
+                        }
+
+                        @Override
+                        public void onDataCompleted() {
+                            try {
+                                if (sPlaylistStreamingObserverImpl != null) {
+                                    sPlaylistStreamingObserverImpl.close();
+                                    sPlaylistStreamingObserverImpl.destroy();
+                                }
+                                playlistService.clearObserverForStreaming();
+                                sSofar++;
+                                Segment newSegment = segmentsQueue.peek();
+                                if (newSegment != null) {
+                                    if (hlsFileDelegate != null) {
+                                        hlsFileDelegate.onProgress(sSofar);
+                                    }
+                                    getHLSFile(context, playlistService, playlistItem,
+                                            segmentsQueue, hlsFileDelegate);
+                                } else {
+                                    if (hlsFileDelegate != null) {
+                                        hlsFileDelegate.onReady(mediaPath);
+                                    }
+                                }
+                            } catch (Exception ex) {
+                                ex.printStackTrace();
+                            }
+                        }
+                    };
+            sPlaylistStreamingObserverImpl =
+                    new PlaylistStreamingObserverImpl(playlistStreamingObserverImplDelegate);
+            playlistService.addObserverForStreaming(sPlaylistStreamingObserverImpl);
+        }
+    }
+
+    private static String getPlaylistIdFromFile(PlaylistItem playlistItem) {
+        String playlistId = "Default";
+        if (playlistItem != null && playlistItem.cached) {
+            String playlistItemMediaPath = playlistItem.mediaPath.url;
+            String[] paths = playlistItem.mediaPath.url.split(File.separator);
+            if (playlistItem.id.equals(paths[paths.length - 2])
+                    && paths[paths.length - 4] != null) {
```

Comment:
what are those `-2` and `-4`? Are we certain that `paths` length is >=4? I would add a comment with an example of how would it look like.